src/main/java/edu/ie3/util/io/FileIOUtils.java

```diff
@@ -141,6 +150,320 @@ public static BufferedWriter getBufferedWriter(
+      throw new FileException("Invalid input file path '" + dirName + "'.", e);
+    }
+
+    /* Open a stream and add content to the archive */
+    try (BufferedOutputStream bufferedOutputStream =
+            new BufferedOutputStream(Files.newOutputStream(validatedOutputFileName));
+        GzipCompressorOutputStream gzipOutputStream =
+            new GzipCompressorOutputStream(bufferedOutputStream);
+        TarArchiveOutputStream tarOutputStream = new TarArchiveOutputStream(gzipOutputStream)) {
+      Files.walkFileTree(
+          source,
+          new SimpleFileVisitor<Path>() {
+            @Override
+            public FileVisitResult visitFile(Path file, BasicFileAttributes attrs)
+                throws IOException {
+              /* Skip symbolic link */
+              if (attrs.isSymbolicLink()) return FileVisitResult.CONTINUE;
+
+              /* Copy files to archive */
+              Path relFile = source.relativize(file);
+              TarArchiveEntry entry = new TarArchiveEntry(file.toFile(), relFile.toString());
+              tarOutputStream.putArchiveEntry(entry);
+              Files.copy(file, tarOutputStream);
+              tarOutputStream.closeArchiveEntry();
+
+              return FileVisitResult.CONTINUE;
+            }
+
+            @Override
+            public FileVisitResult visitFileFailed(Path file, IOException exc) throws IOException {
+              logger.error(
+                  "Unable to add '{}' to '{}'. Stopping compression.",
+                  file,
+                  validatedOutputFileName,
+                  exc);
+              return FileVisitResult.TERMINATE;
+            }
+          });
+
+      /* Close everything properly */
+      tarOutputStream.finish();
+      return true;
+    } catch (IOException e) {
+      throw new FileException("Unable to write to '" + validatedOutputFileName + "'.", e);
+    }
+  }
+
+  /**
+   * Checks, if the output file path actually ends with '.tar.gz' or '.gz'. If not, creates a
+   * filename from the directory or file name. Also, checks if the target file already exists. If it
+   * does not exist, finally creates the file.
+   *
+   * @param dirOrFileName path of the directory or file that should be compressed (full path
+   *     required in case of a directory / full path + filename + file extension required in case of
+   *     a file!)
+   * @param outputFileName the optional output filename (full path + filename + file extension
+   *     required!) if different from the provided filename
+   * @param isDir set to true if trying to compress a directory. Set to false if compressing a
+   *     single file. if set to true, will archive the files before compression
+   * @return Output file
+   * @throws FileException If the archive already exists or unable to create the file
+   */
+  private static File validateOutputFileName(
+      final String dirOrFileName, final String outputFileName, boolean isDir) throws FileException {
```

Comment:
At this place, the toggle is not nice as well, but it fits better in here, so it's okay to leave it as it is.